fix: importing untrusted tls certificates language review (#2406)
codenvy/che-docs: modules/administration-guide/pages/importing-untrusted-tls-certificates.adoc

[id="importing-untrusted-tls-certificates_{context}"]

> = Importing untrusted TLS certificates to {prod-short}

Reviewer: FYI, I can't find this content in the preview build that the PR prompts me to select for previews. https://62d9769557e45f00589af6f8--eclipse-che-docs-pr.netlify.app/docs/next/hosted-che/hosted-che/

Author: Here's the link: https://62d9769557e45f00589af6f8--eclipse-che-docs-pr.netlify.app/docs/next/administration-guide/importing-untrusted-tls-certificates/

By default, external communications between {prod-short} components are encrypted with TLS. Communications of {prod-short} components with external services such as proxies, source code repositories, and identity provider might also require TLS. All communications encrypted with TLS require the use of TLS certificates signed by trusted Certificate Authorities (CA).

> When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, it's necessary to import the CA certificate in the {prod-short} installation so that every {prod-short} component considers them as signed by a trusted CA. Following cases typically require this addition:

Reviewer: User focus. 

Be user and action focused instead of using passive voice. What do you think of this rewrite? 

`When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate in the {prod-short} installation so that every {prod-short} component considers the certificates as signed by a trusted CA. The following cases require trusted CA authorization:` 

The last sentence kind of confused me, what do you think of my suggested rewrite? 

Also, `them` was a little confusing since it could conceivably refer to multiple things. So even though it's wordier, I think for clarity, it's good to say, `considers the certificates as signed by a trusted CA`.

> When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate into the {prod-short} instance so that every {prod-short} component considers the certificates as signed by a trusted CA. Following cases typically require this addition:

Reviewer: This content is good, but it doesn't render properly, at least not in the preview:
<img width="847" alt="Screen Shot 2022-07-22 at 1 59 03 PM" src="https://user-images.githubusercontent.com/70717303/180497159-ab14072f-1b67-4a98-97ac-06fdee22f8bd.png">

Author: fixed

> When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate in the {prod-short} installation so that every {prod-short} component considers the certificates as signed by a trusted CA. Following cases typically require this addition:

Reviewer 2: Installation to me sounds like the process of installing something. This reads ambiguously here.

```suggestion
When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate into the {prod-short} instance so that every {prod-short} component considers the certificates as signed by a trusted CA. Following cases typically require this addition:
```

> * The underlying {orch-name} cluster uses TLS certificates signed by a CA that isn't trusted.

Reviewer: Am I understanding this bullet list right? I'm reading it as a list of cases where you NEED to make sure it's signed by a trusted CA. Therefore, if it is NOT, you must import the trusted CA signature (like you stated above)? 

For minimalism, I think this sentence could be re-phrased, what do you think of this?: 

`The following cases require trusted CA authorization:

- The underlying {orch-name} cluster that uses TLS certificates
- {prod-short} server or workspace components that connect to external services such as {identity-provider} or a Git server that use TLS certificates`

Author: I think it's OK to keep it as it is because the current phrasing takes into account that the components might be using trusted certificates from the get-go.

* {prod-short} server or workspace components connect to external services such as {identity-provider} or a Git server that use TLS certificates signed by an untrusted CA.

{prod-short} uses labeled ConfigMaps in {orch-namespace} as sources for TLS certificates. The ConfigMaps can have an arbitrary number of keys with a random number of certificates each.

[NOTE]
====

> When an OpenShift cluster contains cluster-wide trusted CA certificates added through the link:https://docs.openshift.com/container-platform/4.10/networking/configuring-a-custom-pki.html#nw-proxy-configure-object_configuring-a-custom-pki[cluster-wide-proxy configuration], {prod-short} Operator detects them and automatically injects them into a ConfigMap. {prod-short} automatically labels the ConfigMap with the `config.openshift.io/inject-trusted-cabundle="true"` label. Based on this annotation, OpenShift automatically injects the cluster-wide trusted CA certificates inside the `ca-bundle.crt` key of ConfigMap.

Reviewer: Usability 

I would remove the link. It distracts the user from the task at hand and prompts the user to leave your documentation before even finishing the sentence. 

I think links are best used in prereqs or additional resources. They're too distracting in a procedure or informative paragraph, etc 

`When an OpenShift cluster contains cluster-wide trusted CA certificates added through the cluster-wide-proxy configuration, {prod-short} Operator detects the certificates and automatically injects them into a ConfigMap.`

Reviewer 3: @jc-berger Thank you for going the extra mile on this. I agree that adding the info to Additional resources could be helpful, too. But I don't agree with removing it from the text. Links are an inherent and useful part of the web. It would be a different matter if, let's say, every second term in the paragraph was a link, or if the link was overly long. But that's not the case. This is well placed and useful. Users are well aware of hypertext's capability to include links, and they can use them when they need. It's not confusing.

Btw, see pg. 270 of the IBM SG.

Reviewer: Thanks, Robert :) I always default to the writer knowing their content better than I as the reviewer. I didn't mean to imply links can never be used. As you pointed out, It's a good choice to use here.

====

[IMPORTANT]
====
Some {prod-short} components require a full certificate chain to trust the endpoint.

> If the cluster is configured with an intermediate certificate, add the whole chain (including self-signed root) to {prod-short}.

Reviewer: IBM SG, avoid parenthesis except for acronyms. 

`If the cluster is configured with an intermediate certificate, add the whole chain, including self-signed root, to {prod-short}.`

[id="importing-untrusted-tls-certificates_{context}"]
= Importing untrusted TLS certificates to {prod-short}

By default, external communications between {prod-short} components are encrypted with TLS. Communications of {prod-short} components with external services such as proxies, source code repositories, and identity provider might also require TLS. All communications encrypted with TLS require the use of TLS certificates signed by trusted Certificate Authorities (CA).

> When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate into the {prod-short} instance so that every {prod-short} component considers the certificates as signed by a trusted CA. Following cases typically require this addition:

Reviewer 2: Two suggestions:
* `considers` (sounds like anthropomorphism or AI) > `treats`
* `Following cases typically require this addition` > `You have to do this in the following cases:` (makes it read more user-centric)
```suggestion
When the certificates used by {prod-short} components or by an external service are signed by an untrusted CA, you must import the CA certificate into the {prod-short} instance so that every {prod-short} component treats the certificates as signed by a trusted CA. You have to do this in the following cases:
```

* The underlying {orch-name} cluster uses TLS certificates signed by an untrusted CA.

> * {prod-short} server or workspace components connect to external services such as {identity-provider} or a Git server that use TLS certificates signed by an untrusted CA.

Reviewer 4: Today Keycloak is not the default OIDC provider anymore. Hence we may need to change that in https://github.com/eclipse-che/che-docs/blob/main/antora.yml#L29.

Author: Is there something I can do? Who's the current default provider?

Reviewer 2: @l0rd, what should be the new value for these attributes/variables?
```
    identity-provider-id: keycloak
    identity-provider: Keycloak
```
@deerskindoll, btw, looks like we might need to update the downstream docs as well; currently, this is downstream:
```
:identity-provider-id: keycloak
:identity-provider: RH-SSO
```

Reviewer 2: I propose the following two. They are the same both upstream and downstream, correct? 
```
    identity-provider-id: openshift-oauth
    identity-provider: OpenShift OAuth
```

Reviewer 2: I'm opening a RHDEVDOCS Jira issue for this: https://issues.redhat.com/browse/RHDEVDOCS-4296.

> * {prod-short} server or workspace components connect to external services such as {identity-provider} or a Git server that use TLS certificates signed by an untrusted CA.

Reviewer 2: Thank you, @tolusha, for drafting this change:
```suggestion
{prod-short} server or workspace components connect to external OIDC providers or a Git server that use TLS certificates signed by an untrusted CA.
```

.Procedure

. Save the certificates you need to import to a local file system.
+
[CAUTION]
====

> * A certificate with the introductory phrase `BEGIN TRUSTED CERTIFICATE` is likely in the PEM `TRUSTED CERTIFICATE` format, which is not not supported by Java. Convert it to the supported `CERTIFICATE` format with the following command:

Reviewer: Omit the extra `not`

`...which is not supported by Java.`

$ {orch-cli} create configmap custom-certs --from-file=__<bundle-file-path>__ -n={prod-namespace}
----
+
To apply more than one bundle, add another `-from-file=_<bundle-file-path>_`. Alternatively, create another ConfigMap.

. Label created ConfigMaps with the `app.kubernetes.io/part-of=che.eclipse.org` and `app.kubernetes.io/component=ca-bundle` labels:
+
[subs="+attributes,+quotes"]
----
$ {orch-cli} label configmap custom-certs app.kubernetes.io/part-of=che.eclipse.org app.kubernetes.io/component=ca-bundle -n <{project-context}-namespace-name>
----

. Deploy {prod-short} if it hasn't been deployed before. Otherwise wait until the rollout of {prod-short} components finishes. 
. Restart running workspaces for the changes to take effect.

== Troubleshooting imported certificate issues

> If issues occur after adding the certificates, verify the specified values at a {prod-short} installation and workspace level.

Reviewer 2: Installation to me sounds like the process of installing something. This reads ambiguously here.
```suggestion
If issues occur after adding the certificates, verify the specified values at the {prod-short} instance level and workspace level.
```

> .Verifying imported certificates at the {prod-short} installation level

Reviewer 2: Installation to me sounds like the process of installing something. This reads ambiguously here.

```suggestion
.Verifying imported certificates at the {prod-short} instance level
```

$ {orch-cli} create configmap custom-certs --from-file=__<bundle-file-path>__ -n={prod-namespace}
----
+
To apply more than one bundle, add another `-from-file=_<bundle-file-path>_`. Otherwise, create another ConfigMap. 

> // Wouldn't "alternatively" make more sense than "otherwise"? 

Reviewer: You need to delete this peer review note/question: `// Wouldn't "alternatively" make more sense than "otherwise"?`

. Label created ConfigMaps with the `app.kubernetes.io/part-of=che.eclipse.org` and `app.kubernetes.io/component=ca-bundle` labels:
+
[subs="+attributes,+quotes"]
----
$ {orch-cli} label configmap custom-certs app.kubernetes.io/part-of=che.eclipse.org app.kubernetes.io/component=ca-bundle -n <{project-context}-namespace-name>
----

> . Deploy {prod-short} if it hasn't been deployed before. Otherwise wait until the rollout of {prod-short} components finishes. Restart running workspaces for the changes to take effect.

Reviewer: Avoid contractions and this look like it'd be two commands, not one. 

1. Deploy {prod-short} if it has not been deployed. Otherwise wait until the rollout of {prod-short} components finishes.
2. Restart running workspaces for the changes to take effect.

== Troubleshooting imported certificate issues

> If issues occur after adding the certificates, verify the below specified values at a {prod-short} installation and workspace level.

Reviewer: Accessibility and usability. 

Avoid directional words like `above` and `below`

`If issues occur after adding the certificates, verify the specified values at a {prod-short} installation and workspace level.`

.Verifying imported certificates at the {prod-short} installation level

> * In case of a {prod-short} link:https://docs.openshift.com/container-platform/latest/operators/understanding/olm-what-operators-are.html[Operator] deployment, the namespace where the `CheCluster` is located contains labeled ConfigMaps with the right content:

Reviewer: Again, omit links and the peer review note/question. 

``In case of a {prod-short} Operator deployment, the namespace where the `CheCluster` is located contains labeled ConfigMaps with the correct content:`

Note, I think for word choice `correct` is better since `right` could mean direction, not accuracy.

> // Correct content? Relevant content?

Reviewer: omit this content.

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get cm --selector=app.kubernetes.io/component=ca-bundle,app.kubernetes.io/part-of=che.eclipse.org -n {prod-namespace}
----
+

> Check the content of ConfigMap by entering this command:

Reviewer: `Check the content of ConfigMap by entering:`

OR

`Check the content of ConfigMap by entering the following command:` 

I think the first versionI suggested reads better, but you could rephrase to either one of these :)

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get cm __<name>__ -n {prod-namespace} -o yaml
----

* {prod-short} Pod Volumes list contains a volume that uses `ca-certs-merged` ConfigMap as data-source.
To get the list of Volumes of the {prod-short} Pod, run:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pod -o json __<{prod-id-short}-pod-name>__ -n {prod-namespace} | jq .spec.volumes
----
+

> * {prod-short} mounts certificates in the `/public-certs/` folder of the {prod-short} server container. Enter the following command to view the list of files in this folder:

Reviewer: Reword to be consistent with your previous commands: 

`To view the list of files in this folder, enter:`

+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} exec -t __<{prod-id-short}-pod-name>__ -n {prod-namespace} -- ls /public-certs/
----
+

> * In the {prod-short} server logs, there is a line for every certificate added to the Java truststore, including configured {prod-short} certificates. View them with following command:

Reviewer: For all these bullet points, you precede the action with context. I'm not used to seeing this. 

Can you simply give the users a series of actions and make this a procedure module?

Author: this will be addressed in https://issues.redhat.com/browse/RHDEVDOCS-4246

(...)
----
+
To get the SHA1 hash of a certificate on the local filesystem, run:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ openssl x509 -in __<certificate-file-path>__ -fingerprint -noout
SHA1 Fingerprint=3F:DA:BF:E7:A7:A7:90:62:CA:CF:C7:55:0E:1D:7D:05:16:7D:45:60
----

.Verifying imported certificates at the workspace level

* Start a workspace, obtain the {orch-namespace} name in which it has been created and wait for the workspace to be started.

* Get the name of the workspace Pod:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ {orch-cli} get pods -o=jsonpath='{.items[0].metadata.name}' -n __<workspace namespace>__ | grep '^workspace.*'
----

> * Get the name of the Che-Theia IDE container in the workspace Pod:

Reviewer 4: Che-theia is just one of the IDEs that we support so we may want to change that, probably in another PR.

(...)
----
+
To get the SHA1 hash of a certificate on the local filesystem, run:
+
[subs="+attributes,+quotes",options="nowrap",role=white-space-pre]
----
$ openssl x509 -in __<certificate-file-path>__ -fingerprint -noout
SHA1 Fingerprint=3F:DA:BF:E7:A7:A7:90:62:CA:CF:C7:55:0E:1D:7D:05:16:7D:45:60
----

.Verifying imported certificates at the workspace level

> * Start a workspace, obtain the {orch-namespace} name in which it has been created, and wait for the workspace to be started.

Reviewer: No need for the comma before `and` 

`...it has been created and wait for the workspace to be started.`

> // What do I need the name for?

Reviewer: remove this content

> * Get the name of the workspace Pod with the following command:

Reviewer: minimalism 

Avoid the whole `with this command` it fluff and unnecessary.
